### DISCUSSION TITLE
How to launch a WPF application using a custom native host

### DISCUSSION THREAD

**Discussion by Rashmatash** [dotnet/runtime#85421] [Q&A] [answered]
Hi,

Looking at a WPF/.Net 6 application, I see that there's an exe-file which is the host/launcher of the actual .Net application, which is in a dll-file. Is it possible to replace this .exe file with my own using https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting?
I tried it and although I managed to get a function pointer to a user-defined method in `App` which then calls `Main`, it always crashes without giving any error messages. Does anyone know how I should go about doing this?

Basically I just want to do this: `dotnet MyApp.dll`, but instead of dotnet, I'd like to use my own C++ launcher.

**Comment by vitek-karas** [comment, marked as answer]
Use the `hostfxr_run_app` instead of getting a function pointer to a method.
Some more description of the available functionality is here: https://github.com/dotnet/runtime/blob/main/docs/design/features/native-hosting.md

**Reply by vitek-karas** [reply]
One more note: The `exe` which is generated by the build will include Win32 resources from the app, so things like the manifest and icon. If you use your custom host there might be some incompatibilities with the app - especially the manifest can be important in some cases.

**Reply by Rashmatash** [reply]
Good, point. I'll try and add anything that's missing if I can. Thanks for your answer.

The reason I wanted to write my own host is so that I can export symbols like `extern "C" {extern __declspec(dllexport) in x = 123;}` for D3D12 loader would pick it up and use the Agility SDK. Do you happen to know a better way of exporting symbols from the main exe in the process?

**Reply by Rashmatash** [reply]
BTW, I used `hostfxr_main_startupinfo_fn`. Is that the same thing or does `hostfxr_run_app` has more advantages? The documentation states that the context for `hostfxr_run_app` should come from `hostfxr_initialize_for_dotnet_command_line`. Would that work with a WPF application since that's not a command line app?

**Reply by vitek-karas** [reply]
`hostfxr_main_startupinfo_fn` has lot of backward compat stuff in it - to meet what dotnet.exe has been doing. If you only want to run an app then `hostfxr_run_app` should be simpler.

> The documentation states that the context for hostfxr_run_app should come from hostfxr_initialize_for_dotnet_command_line

The "dotnet command line" in this case refers to the `dotnet.exe`. It doesn't refer to console versus GUI apps. It should work for both (with the caveat of the Win32 resources mentioned above).

**Comment by Rashmatash** [comment]
I looked at the source for dotnet.exe and was able to write my own host.
